src/storage/file_storage.rs

```diff
@@ -0,0 +1,274 @@
+use std::{
+    collections::{HashMap, VecDeque},
+    fs::{self, File},
+    io::{self, BufRead},
+    path::PathBuf,
+    time::SystemTime,
+};
+
+use crate::Error;
+
+pub struct FileWriterFactory {
+    writer_type: FileWriterType,
+    use_test_dir: Option<fn() -> Result<PathBuf, Error>>,
+}
+
+impl FileWriterFactory {
+    pub fn new(writer_type: FileWriterType) -> Self {
+        Self {
+            writer_type,
+            use_test_dir: None,
+        }
+    }
+
+    pub fn testing_with(mut self, test_dir: fn() -> Result<PathBuf, Error>) -> Self {
+        self.use_test_dir = Some(test_dir);
+        self
+    }
```

Comment:
This whole use_test_dir and testing_with seems dubious. once we simplify these types I think this will be way easier to just pass in a prefix dir as a parameter to FileKeyStorage? especially once the interactor things go away and we use a more direct style:

```rust
impl FileKeyStorage {
  fn new(base_path: PathBuf) -> Self {
     Self { base_path }
  }

  fn get_keys_internal(&self) -> Result<Vec<Keypair>, KeyStorageError> {
    let path = self.base_path.join(DataPaths::Keys.path());
    let directory = Directory::new(path)
    // etc
  }
}
```